### DISCUSSION TITLE
Extension / Plug-in Support

### DISCUSSION THREAD

**Discussion by kyaso** [Zettlr/Zettlr#3108] [Ideas]
I know that this is a big topic. But definitely one worth discussing. As Zettlr's userbase is constantly growing, so does the demand for new features. Unfortunately, not every feature request can make it into the app in the long run.

This is where extensions come in. They would allow anyone to easily customize the behavior of Zettlr to fit their needs (+ share it with others). Other benefits:

- Zettlr's codebase stays clean, and only necessary features get implemented
- Users can select only the extensions they want for their particular workflow
- More people get encouraged to contribute new functionality to Zettlr (developing extensions could be more accessible than fiddling with the main code)
- Extension could allow "in-field" testing of new features that didn't make it into the codebase. If the extension turns out to be popular, its functionality can be integrated into Zettlr directly.

A few aspects that came into my mind regarding the implementation of an extension infrastructure:
- Extension API
- Extension Marketplace
- Extension Security

I might have missed some aspects, and I also don't know the effort required for this endeavor. That's why I thought it might be a good idea to open a discussion. So that we have a central place to talk about this topic (maybe there is already one?).

**Comment by nathanlesage** [comment]
@tobiasdiez has had some very good points on plugins yes/no in the past, and I've generally stuck to his advice in this regard. Feel free to add some of your points here again, or link them, if you think that this is advisable, Tobias.

_However_, that being said: I'm slowly starting to see that JabRef and Zettlr differ vastly in functionality. JabRef is a reference manager, and thus there is only so much that needs to be done. But over the past four years developing Zettlr, I have slowly but steadily come to the conclusion that writing text is much more complex and requires a lot of different tools, and many people have vastly differing requirements. I think reference managers can only do that much (and that is why I stick with only two plugins for Zotero, because everything else is simply not in the scope of a reference manager), but a writing app should offer much more stuff.

That being said: Writing a plugin system is incredibly difficult, and thus we should do that extremely limited. Two very good suggestions for "trying this out" would be (a) a renderer plugin API, so that people can render whichever elements they like within the app, and (b) a language server API so that people can incorporate external services for linting code, spellchecking, and what not else.

That might keep the effort manageable at least for now. New features to the plugin API should only be incorporated after an insane amount of deliberation.

With regard to security, there may be an easy way: invisible Browser windows. Each plugin could run in its own dedicated browser window. This would ensure that it's a fully autonomous process and can only interact within that sandbox. There would be just two things we need to remember: First, the `broadcastIPCMessage` function would need to be scoped not to accidentally leak data into the plugin windows, and we would need extremely strong validation within the preload for these plugin windows. Also, we should prevent any network access for these. (Alternatively, because I already see people complaining that some language service requires an API access, we could implement a rights management so that the user is prompted something like "Plugin xyz wants to access website abc, allow?" — but here you can already see how incredibly difficult that could be)

So, it's worth a discussion, but we need to tread carefully.

**Reply by tobiasdiez** [reply]
I agree that my experience with JabRef cannot directly be translated to Zettlr. Still the two major points of problems we had were:
- At some points plugins are no longer actively maintained, which then gives also a bad reputation for the original app.
- Development of the main app gets slower as one has to make sure that the plugins don't break. Also testing that everything is still working is a manual process.

These issues would also apply to Zettlr I think. One should also keep in mind that Zettlr's target group is more "non-technical" writers, and those normally don't have the knowledge to code plugins. Thus while for example the Zettlr community is larger, the pool of potential plugin writers is not necessary so.

In my opinion, it make sense to have plugins if the community gets so big that they would for example jump in maintaining plugins if the original maintainer stopped doing so. For now I would advice to make it easier to contribute to Zettlr itself. Once @nathanlesage gets so many PRs about functionality that he actually doesn't want to have in the main repo, then I would say it's a good point to think about plugins. At the beginning they could also be just a collection of classes in a "plugins" folder in the main repo and giving users the possibility to switch them on/off. In this way one doesn't have to worry about extension store, security and backwards compatibility.

**Reply by nathanlesage** [reply]
I agree with your points.

> At some points plugins are no longer actively maintained, which then gives also a bad reputation for the original app.

One thing we could do to both ensure proper maintenance and prevent a bad reputation would be to enforce upper version constraints. As soon as we upgrade the plugin API, we could simply disable too old plugins and give out a notice. As long as the plugin API moves slower than the main app, this could be a compromise.

> Development of the main app gets slower as one has to make sure that the plugins don't break. Also testing that everything is still working is a manual process.

That's something that I would argue is already handled by having the Discord. It's easy to just drop a deprecation note there, and anyone developing plugins could look into that specific channel. As long as we're transparent that we're not taking hostages that might ensure both that plugins only get developed by people who want to commit to maintaining it.

However, I see that it's a big undertaking. Therefore, I like the approach of limited plugins (only rendering or only linting for now). This way we could very gently introduce a plugin API.

> Once @nathanlesage gets so many PRs about functionality that he actually doesn't want to have in the main repo, then I would say it's a good point to think about plugins.

I would argue when the point comes that I'm overwhelmed by PRs it's way too late to start thinking about Plugins.

**Comment by deletosh** [comment]
Those said, is there a roadmap for limited plugin support?

**Reply by huyz** [reply]
At this point, I wonder how much it would make sense to offer a certain level of Obsidian API compatibility, just so that it's easier for some of the large number of Obsidian plugins to be ported…

**Reply by espipj** [reply]
Obsidian API compatibility would be super great!

**Reply by AltayYuzeir** [reply]
Especially for something like the BMO-Chatbot

**Reply by DutchPete** [reply]
@AltayYuzeir among the now 1000s of Obsidian plug-ins there are many that would qualify for porting.

**Comment by inspiredearth** [comment]
Interesting discussion. After reading the pros and cons, and related thoughts, shared here, what comes to mind is a question. Obsidian seems to successfully implement a plug-in system, with a thriving community developing, maintaining and using the vast array of functionality these plug-ins provide.

NotePlan (another MarkDown based knowledge, _and_ task etc. management tool I use) also has a well established and excellent plug-in system.

So I wonder ... is the notion of whether or not it's a good idea, whether the cons outweigh the pros, a case of over thinking this whole idea?
In other areas, tools like WordPress, browsers (Chrome, Firefox, etc.) have all proven that allowing for a plug-in frame work is a useful and self-maintaining in practice. The wider community will deal with the ups and downs of whether or not a given plugin is maintained long term or not, for example. That does not need to be Zettlr's concern.

I see no fundamental or overbearing disadvantage to allowing Zittlr to have expandable functionality through plugins. Other than the burden of having to significantly modify the underlying codebase to accommodate plugins. That will be a sizable task. But other than deciding if that's a task there is the time and resource to tackle, the wide questions of whether or not whether making Zettlr extension-capable is a good idea or not seems like a mute point to me.

**Comment by DutchPete** [comment]
Is this topic still under consideration or has it died a quiet death?

**Reply by nathanlesage** [reply]
That topic always simmers. It's a very difficult one and before I have a good strategy to tackle this securely and with a manageable amount of maintenance I won't implement it, but it's not that I'm carrying a hard no. But again, don't get your hopes up; this is still a free time project, and currently free time is very sparse.

**Reply by DutchPete** [reply]
I understand, you will decide if and when this is taken any further.

**Reply by Orphist** [reply]
As another way to impletemnt plugins system is Joplin (like Obsidian)
https://github.com/laurent22/joplin

its official plugin repo:
https://github.com/joplin/plugins

**Reply by DutchPete** [reply]
> As another way to impletemnt plugins system is Joplin (like Obsidian) https://github.com/laurent22/joplin
> 
> its official plugin repo: https://github.com/joplin/plugins

I tried Joplin, and although it has a lot of positive aspects, the one thing I dislike and ultimately was a deal breaker for me is the fact that notes are stored in an SQLite database. So, if ever you want to get hold of them you have to export them 1st.

I prefer the Zettlr and Obsidian way whereby your notes are stored as notes so they are accessible from the system file explorer.

**Reply by Ehesh** [reply]
> the one thing I dislike and ultimately was a deal breaker for me is the fact that notes are stored in an SQLite database. 

I did not know that. I've been looking for a light pkm for the past week, but like really suffering cause I wanted something lightweight and customizable.  I was considering going to joplin but did not know you'd need to go though sqlite before accessing the files, bummer. And I got excited when I read a post in github soewhere that the author was thinking of porting zettlr to Tauri (this post was about 2 years ago). Electron really bloats the ram usage, Obsidian takes about 400 mb ram, (havent reinstalled zettlr in a while so cant compare), but if zettlr is not in tauri or anything lighter that electron and does not have the option for pluggins... well.... I'm not loving it...

**Reply by DutchPete** [reply]
Another option is QOwnNotes, which is customizable. But what puts me off, and it did for Joplin too, is the split panes for editing and reading modes. Neither Obsidian nor Zettlr have it, and that is such a nice thing. Typora is like Obs and Zettlr, but is not free.

**Reply by Ehesh** [reply]
> is the split panes for editing and reading modes. 

are  you refering to the WYSIWYG (What You See Is What You Get) that automatically renders the text in _"pretty"_  format? 
I did investigate QOwnNotes and did not like that part, it lacks WYSIWYG. Jopin does have it tho. just like obsidian, it renders the text as you type it. You do have the option of viewing the file in `source` mode which removes the fancy eyecandy and displays as just text (obsidian also does this BTW, in fact obsidian has 3 modes, source, live preview, and reading). 
I considered going back to emacs cause it would be about 50-100 mb ram MAX. But when using markdown you dont get WYSIWYG. there is something similar to that in orgmode and I think orgdown, but the syntax is a bit diferent. I dont mind that part, what I do mind is that it lacks custom CSS styling. As emacs is a text editor, I cant have stuff like infoboxes like in obsidian, or buttons, or dropdowns or forms like I can with some of the pluggins in obsidian. Of course there is emacs widget library but it seems it does not run as good.

**Reply by DutchPete** [reply]
I disagree about Joplin. Having said that, I don't want go farther down this path because I don't think it is fair to discuss other apps on this Zettlr page.

**Comment by lordofscripts** [comment]
Plugins would be nice. A plugin I would be interested in having or even doing depending on the development platform, is one that could encrypt my text file into something that remains as text rather than binary. Simplest way would be like ROT13 (text results from text) but much better.
